docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,331 @@
+```
+
+Since i-tokens are constantly accruing interest, they do not have the same
+properties as normal tokens. Most importantly, i-tokens held in two different
+accounts are only fungible if they have been updated to the same time. For example,
+10 i-tokens in the year 1999 are worth much more than 10 i-tokens in the year 3000, since
+those tokens from 1999 have over 1,000 years of interest (see Futurama). Because
+of this, we have to compound interest before any minting, burning, and
+transferring.
+
+### Initial Attempt
+
+The concept seems simple to implement. We can create an SPL token, allow
+tokens to be minted by a new "interest accruing" program, and then make sure to
+call it regularly.
+
+This solution has a few problems. First, how do we compound interest regularly?
+We can create a separate service which periodically sends transactions to update
+all of the accounts. Unfortunately, this will cost the service a lot, especially if
+there are thousands of accounts to update.
+
+We can enforce compounding before any transfer / mint / burn, but then all programs
+and web consumers need to change to make these update calls. How do we change programs
+and wallets to properly update accounts without creating a big `if is_i_token` case
+all over the code?
+
+This solution creates friction for almost all SPL token users: on-chain programs,
+wallets, applications, and end-users.
+
+## Proposed Solution
+
+Following the approach taken by Aave, we can lean on a crucial feature of i-tokens:
+although the token holding amounts are constantly changing, each token holder's
+proportion of total supply stays the same. Therefore, we can satisfy the
+fungibility requirement for tokens on proportional ownership. The actual token
+amounts used for token transfers / mints / burns are just an interface, and
+everything is converted to proportional shares inside the program.
+
+Let's go through an example. Imagine we have an i-token mint with a total of
+10,000 proportional shares, a 20% annual interest rate, and 1 share is currently
+worth 100 tokens. If I transfer 1,000 tokens today, the program converts that and moves
+10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
+1,000 tokens, the program converts and only moves 8.33 shares. The same concept
+applies to mints and burns.
+
+Following this model, we never have to worry about compounding interest. We
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+since i-tokens need to be easily integrated in on-chain programs, web apps,
+and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
```

Comment:
Also, for further easiness of creating interesting i-tokens, I think we should provide some scaffolding mechanism by rust.

maybe like this?:

```
trait InterestBearingToken {
  fn token_amount_to_share_amount() -> u64,
  fn share_amount_to_token_amount() -> u64,
}

trait Token<T: InterestBearingToken> {
  fn process_transfer_checked(...) {
     ...
     self.process_transfer_checked_internal(T::token_to_share(amount));
     ...
  }
  fn process_get_holding_balance(...) {
    ....
     return T::share_to_token(self.amount)
  }
}
``` 

As a testbet, I think we can refactor current existing implementation like this?

The intention here is that avoiding the plain copy-paste; because this will be nightmere for patching.
So, itokens developer only need to care about their cool present-value/recoup/interest/etc quant calculation. :)
Then, pretty basic invariant of no phantom tokens and authorization-checking code (this is pretty battle-tested now) can be reused.

Reply:
Very good idea, we should absolutely re-use as much of the existing SPL token code as possible